File: web-react/src/services/authService.js
import axios from 'axios';

// Định nghĩa API endpoint với cổng
const API_URL = 'http://localhost:3000/api/auth'; // Cập nhật URL với cổng

// Hàm gọi API để đăng nhập
export const login = async (account, password) => {
  try {
    // Xác định xem tài khoản là email hay số điện thoại
    const isEmail = /\S+@\S+\.\S+/.test(account);
    
    const response = await axios.post(`${API_URL}/login`, 
      { 
        [isEmail ? 'email' : 'phone']: account,
        password 
      }, 
      { headers: { 'Content-Type': 'application/json' } }
    );
    const { accessToken, user } = response.data; // Lấy token và thông tin người dùng từ phản hồi
    return { accessToken, user };
  } catch (error) {
    const errorMessage = error.response?.data?.message || 'Đăng nhập thất bại';
    throw new Error(errorMessage); // Ném lỗi với thông điệp cụ thể
  }
};

// Hàm gọi API để đăng ký
export const register = async (userData) => {
  try {
    const response = await axios.post(`${API_URL}/register`, userData);
    return response.data; // Trả về dữ liệu từ server
  } catch (error) {
    // Xử lý lỗi và trả về thông báo phù hợp
    console.error('Registration error:', error.response?.data); // Log chi tiết lỗi
    const errorMessage = error.response?.data?.message || 'Đăng ký thất bại';
    throw new Error(errorMessage); // Ném lỗi với thông điệp cụ thể
  }
};

export const verifyCode = async (email, verificationCode) => {
  try {
    const response = await axios.post(`${API_URL}/verify`, {
      email, // Bao gồm email trong yêu cầu
      verificationCode, // Bao gồm mã xác thực
    });
    return response.data; // Trả về dữ liệu từ server
  } catch (error) {
    // Xử lý lỗi và trả về thông báo phù hợp
    console.error('Verification error:', error.response?.data); // Log chi tiết lỗi
    const errorMessage = error.response?.data?.message || 'Xác thực thất bại';
    throw new Error(errorMessage); // Ném lỗi với thông điệp cụ thể
  }
};

export const resendVerificationCode = async (email) => {
  try {
    const response = await axios.post(`${API_URL}/resend`, {
      email, // Gửi email để gửi lại mã
    });
    return response.data; // Trả về phản hồi từ server
  } catch (error) {
    console.error('Resend code error:', error.response?.data); // Log lỗi
    const errorMessage = error.response?.data?.message || 'Gửi mã xác thực thất bại';
    throw new Error(errorMessage); // Ném lỗi với thông điệp cụ thể
  }
};

File: app_flutter/lib/services/AuthService.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:connectivity_plus/connectivity_plus.dart';
import "package:jwt_decode/jwt_decode.dart";

class AuthService {
  final String baseUrl = 'http://192.168.100.243:3000/api/auth';
  final storage = FlutterSecureStorage();

  Future<void> register(
      String name, String email, String password, String phone) async {
    final url = Uri.parse('$baseUrl/register');
    final connectivityResult = await Connectivity().checkConnectivity();

    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'name': name,
          'email': email,
          'password': password,
          'phone': phone,
        }),
      );

      if (response.statusCode == 201) {
        await _saveUserEmail(email);
      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Đăng ký thất bại: $e');
    }
  }

  Future<void> verifyEmail(String verificationCode) async {
    final email = await _getUserEmail();

    if (email == null) {
      throw Exception("Không có email để xác thực");
    }

    final url = Uri.parse('$baseUrl/verify');
    final connectivityResult = await Connectivity().checkConnectivity();

    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'verificationCode': verificationCode,
          'email': email,
        }),
      );

      if (response.statusCode == 200) {
        await _clearUserEmail();
      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Xác thực thất bại: $e');
    }
  }

  Future<void> resendVerificationCode() async {
    final email = await _getUserEmail();
    if (email == null) {
      throw Exception("Không có email để gửi lại mã xác thực");
    }

    final url = Uri.parse('$baseUrl/resend');
    final connectivityResult = await Connectivity().checkConnectivity();

    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email}),
      );

      if (response.statusCode == 200) {

      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Gửi lại mã xác thực thất bại: $e');
    }
  }

  Future<void> login(String emailOrPhone, String password) async {
    final connectivityResult = await Connectivity().checkConnectivity();
    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }
    if (emailOrPhone.isEmpty) {
      throw Exception('Email hoặc số điện thoại phải được nhập');
    }
    if (password.isEmpty) {
      throw Exception('Mật khẩu không được để trống');
    }
    final url = Uri.parse('$baseUrl/login');

    try {
      final bodyData = {
        'password': password,
        if (_isValidEmail(emailOrPhone))
          'email': emailOrPhone
        else
          'phone': emailOrPhone
      };

      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(bodyData),
      );

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);

        if (responseData != null &&
            responseData['_id'] != null) {
          final String userId = responseData['_id'];
          final String token = responseData['accessToken'];

          final prefs = await SharedPreferences.getInstance();
          await prefs.setString('accessToken', token);
          await prefs.setString('userId', userId);

          await _saveToken(token);
        } else {
          throw Exception('Phản hồi không chứa thông tin người dùng hợp lệ.');
        }
      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Đăng nhập thất bại: $e');
    }
  }

  bool _isValidEmail(String input) {
    final emailRegex = RegExp(r"^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$");
    return emailRegex.hasMatch(input);
  }

  Future<void> forgotPassword(String email) async {
    final url = Uri.parse('$baseUrl/forgot-password');
    final connectivityResult = await Connectivity().checkConnectivity();

    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email}),
      );

      if (response.statusCode == 200) {
        await _saveUserEmail(email);
      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Quên mật khẩu thất bại: $e');
    }
  }

  Future<void> resetPassword(
      String verificationCode, String newPassword) async {
    final email = await _getUserEmail();
    if (email == null) {
      throw Exception("Không có email để đặt lại mật khẩu");
    }

    final url = Uri.parse('$baseUrl/reset-password');
    final connectivityResult = await Connectivity().checkConnectivity();

    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'verificationCode': verificationCode,
          'password': newPassword,
        }),
      );

      if (response.statusCode == 200) {
        await _clearUserEmail();
      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Đặt lại mật khẩu thất bại: $e');
    }
  }

  Future<void> _saveToken(String token) async {
    await storage.write(key: 'authToken', value: token);
  }

  Future<void> _saveUserEmail(String email) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('email', email);
  }

  Future<String?> _getUserEmail() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('email');
  }

  Future<void> _clearUserEmail() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('email');
  }

  Future<String?> getToken() async {
    return await storage.read(key: 'authToken');
  }

  Future<void> logout() async {
    await storage.delete(key: 'authToken');
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('accessToken');
    await prefs.remove('userId');
  }

  Future<void> refreshToken() async {
    final url = Uri.parse('$baseUrl/refresh-token');
    final connectivityResult = await Connectivity().checkConnectivity();

    if (connectivityResult == ConnectivityResult.none) {
      throw Exception('Không có kết nối mạng');
    }

    try {
      final response = await http.post(
        url,
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        final String newToken = responseData['accessToken'];

        await _saveToken(newToken);

      } else {
        _handleError(response);
      }
    } catch (e) {
      throw Exception('Làm mới token thất bại: $e');
    }

  }

  Future<bool> checkTokenExpiration() async {
    final token = await getToken();
    if (token == null) {
      throw Exception('Không tìm thấy token');
    }

    bool isExpired = Jwt.isExpired(token);

    if (isExpired) {
      try {
        await refreshToken();
        return true; // Token was expired, but successfully refreshed
      } catch (e) {
        await logout();
        throw Exception('Phiên đăng nhập đã hết hạn, vui lòng đăng nhập lại.');
      }
    }

    return true;
  }


  void _handleError(http.Response response) {
    final errorData = jsonDecode(response.body);
    final message = errorData['message'] ?? 'Lỗi không xác định';

    switch (response.statusCode) {
      case 400:
        throw Exception('Yêu cầu không hợp lệ: $message');
      case 401:
        throw Exception('Không có quyền truy cập: $message');
      case 500:
        throw Exception('Lỗi từ server: $message');
      default:
        throw Exception('Lỗi: $message');
    }
  }
}
